### ISSUE TITLE
Handle promises inside mapWith TransformationType callback

### ISSUE THREAD

**Issue by manuelnucci** [nartc/mapper#185]
## I'm submitting a...
<pre><code>
[ ] Regression
[ ] Bug report
[x] Feature request
[ ] Documentation issue or request
[ ] Support request
</code></pre>

## Current behavior
When mapping a property wrapped in a promise, mapper cannot retrieve the actual value of it. The only workaround that I've found is implement a mapWith transformation inside Profile and manage promise there. However, as the callback expects to return an actual value, is completely impossible to await promise and return its result.

## Expected behavior
Callback inside mapWith (and all others transformation type functions) could expect to receive a raw value or a promise and await its result. This way it could be possible to have lazy loading of entities and, at the same time, map those properties correctly.

## Minimal reproduction of the problem with instructions
Here's a Stack Overflow question where I described the problem: https://stackoverflow.com/questions/63439664/how-to-handle-promises-in-mapwith-transformationtype-nartc-automapper

## What is the motivation / use case for changing the behavior?
In combination with NestJS and TypeORM, main goal would be to benefit from lazy loading and at the same time map properties to their corresponding nested models.

## Environment

<pre><code>
"nestjsx-automapper": "^3.0.23"
"@nartc/automapper-transformer-plugin": "^1.0.20"
 
- Node version: v12.18.3
- Platform:  Windows
</code></pre>

**Comment by nartc** [comment]
Quick work around would be to use an `afterMap` callback. I am going to copy what you have in SO:

```ts
@Entity('careers')
export class Career {
  @PrimaryGeneratedColumn('uuid')
  id!: string;

  @Index('name-idx', { unique: true })
  @Column()
  name!: string;

  @AutoMap(() => CareerCourseRelation)
  @OneToMany(() => CareerCourseRelation, (careerCourseRelation) => careerCourseRelation.career)
  careerCourseRelations!: Promise<CareerCourseRelation[]>;
}
```

```ts
@Exclude()
export class ResponseCareerDto {
  @Expose()
  id!: string;

  @Expose()
  name!: string;

  @Expose()
  @AutoMap(() => ResponseTernaryDto) // <- DTO of CareerCourseRelation
  careerCourseRelations!: ResponseTernaryDto[];
}
```

We're going to focus on the `Profile`

```ts
@Profile()
export class CareerProfile extends ProfileBase {
  constructor(mapper: AutoMapper) {
    super();
    mapper.createMap(Career, ResponseCareerDto)
      .forMember(
         (dest) => dest.careerCourseRelations,
         ignore() // first, we will ignore it from the mapping operations
       )
      .afterMap(async (source, destination) => {
         const resolvedValues = await source.careerCourseRelations;
         destination.careerCourseRelations = mapper.mapArray(resolvedValues, ResponseTernaryDto, CareerCourseRelation);
      });
  }
}
```

However, since `afterMap` is now an async callback, it will run at the end of the current Event Loop, you will need to use the `mapAsync` version of `AutoMapper`.

```ts
// before
this.mapper.map(career, ResponseCareerDto, Career);
```

**Comment by manuelnucci** [comment]
Sorry it took me some time to reply. I was making tests with the code you provided. As expected, works great and allows me to handle promises inside profile class.

Nevertheless, my initial problem arose from the fact that I wanted to use lazy loading as much as I could inside a NestJS app using TypeORM. As I continued digging a bit further in the documentation and in some issues, I see now that it's not a "well" practice to have Promises around, it seems is better to deliver that logic to the `QueryBuilder` or `find*` methods. Moreover, when using Promises to trigger lazy loading as necessary, some of the Profiles that I imported would blow up the stack and crash the app. It seems that the mapper doesn't support Promises natively.

Having stated the facts, I believe I'll stick to depend on eager loading or just requesting the related entities when I need them. That way I can keep using automapper and TypeORM together.

If any of what I've just said doesn't seem to be right, please let me know.

Here're the links which I read to arrive to my conclusions:
- https://typeorm.io/#/eager-and-lazy-relations (last paragraph)
- https://github.com/typeorm/typeorm/issues/3703
- https://github.com/typeorm/typeorm/issues/2315#issuecomment-398351133

Lastly, in reference to `mapDefer`, I had consulted the docs before because, as you said, the name implies to have something related with async. It could be a good place to handle promises. Just having to ignore a property to later retrieve and transform it seems to be a bit overkill IMO. I'll leave it to you if this feature is really worth it.

**Comment by nartc** [comment]
@manuelnucci you are correct. `AutoMapper` does not support Promises natively. One reason is that `AutoMapper` because mapping configurations are pretty static and usually are constructed (and should be ready) before any of the app's operations. 

With that said, I will still consider adding Promises support to `mapDefer` because as you said, you should be able to configure a transformation on a `destination.member` from an asynchronous `source.member`. And also, `mapDefer` would be the only Map Function that supports Promises. All of the other map functions, it's either weird/incomplete to support some and not the others. So I decide to not support any other map functions to have Promises except for MapDefer.

**Comment by manuelnucci** [comment]
Great to know. Agree with you that it should be implemented in a schema all or one. Thanks for the notice.

**Comment by manuelnucci** [comment]
@nartc Ok, thanks for the notice. I'll stick to use `afterMap` for now. Hope to use as less promises as I can then.

**Comment by marcoacevey** [comment]
> Sorry it took me some time to reply. I was making tests with the code you provided. As expected, works great and allows me to handle promises inside profile class.
> 
> Nevertheless, my initial problem arose from the fact that I wanted to use lazy loading as much as I could inside a NestJS app using TypeORM. As I continued digging a bit further in the documentation and in some issues, I see now that it's not a "well" practice to have Promises around, it seems is better to deliver that logic to the `QueryBuilder` or `find*` methods. Moreover, when using Promises to trigger lazy loading as necessary, some of the Profiles that I imported would blow up the stack and crash the app. It seems that the mapper doesn't support Promises natively.
> 
> Having stated the facts, I believe I'll stick to depend on eager loading or just requesting the related entities when I need them. That way I can keep using automapper and TypeORM together.
> 
> If any of what I've just said doesn't seem to be right, please let me know.
> 
> Here're the links which I read to arrive to my conclusions:
> 
> * https://typeorm.io/#/eager-and-lazy-relations (last paragraph)
> * [Please clarify: Is lazy-loading with promises supported or not? typeorm/typeorm#3703](https://github.com/typeorm/typeorm/issues/3703)
> * [Lazy loading with eager loading in combination doesn't work typeorm/typeorm#2315 (comment)](https://github.com/typeorm/typeorm/issues/2315#issuecomment-398351133)
> 
> Lastly, in reference to `mapDefer`, I had consulted the docs before because, as you said, the name implies to have something related with async. It could be a good place to handle promises. Just having to ignore a property to later retrieve and transform it seems to be a bit overkill IMO. I'll leave it to you if this feature is really worth it.

I followed you considerations and im an using eager relations, considering that we gonna load the relations values anyway to map it on to the DTO, lazy loading dont make to mutch sense than,
